File: adapters/multisim/codec.mjs
import { createRequire } from "node:module";
import { resolve } from "node:path";
import { readFileSync, statSync, writeFileSync } from "node:fs";

try {
  const [mode, input, output, packageRoot] = process.argv.slice(2);
  if (!["decode", "encode"].includes(mode)) throw new Error("Invalid codec operation");
  if (statSync(input).size > 32 * 1024 * 1024) throw new Error("Native input exceeds 32 MiB");
  const source = readFileSync(input);
  const require = createRequire(resolve(packageRoot, "package.json"));
  const { decodeBuffer, encodeBuffer, formatByKey } = require("electronics-workbench-decoder");
  let result;
  if (mode === "decode") {
    const signature = Buffer.from("MSMCompressedElectronicsWorkbenchXML");
    if (source.length < 44 || !source.subarray(0, signature.length).equals(signature)) {
      throw new Error("Unsupported Multisim container");
    }
    if (source.readBigUInt64LE(signature.length) > 64n * 1024n * 1024n) {
      throw new Error("Decoded native document exceeds 64 MiB");
    }
    result = Buffer.from(decodeBuffer(source).xml);
  } else {
    result = Buffer.from(encodeBuffer(source, formatByKey("multisim")));
    if (!Buffer.from(decodeBuffer(result).xml).equals(source)) throw new Error("Codec round trip failed");
  }
  writeFileSync(output, result, { flag: "wx" });
  console.log(JSON.stringify({ success: true, bytes: result.length }));
} catch (error) {
  console.error(String(error));
  process.exitCode = 1;
}

File: adapters/multisim/codec.py
"""Bounded wrapper around the pinned Electronics Workbench codec."""

from __future__ import annotations

import os
import shutil
import subprocess
from pathlib import Path
from tempfile import TemporaryDirectory

from lxml import etree

from .xmlio import native_xml_bytes, parse_native_xml


class NativeCodec:
    def __init__(self, home: Path | None = None) -> None:
        self.home = home or Path(
            os.environ.get("BIFROST_AUTHORING_HOME", Path.home() / ".bifrost-authoring")
        )
        self.node = shutil.which("node")
        if (
            not self.node
            or not (self.home / "node" / "node_modules" / "electronics-workbench-decoder").is_dir()
        ):
            raise ValueError(
                "Native authoring codec is not configured; run the authoring setup command."
            )

    def _call(self, mode: str, source: Path, destination: Path) -> None:
        result = subprocess.run(
            [
                self.node,
                str(Path(__file__).with_name("codec.mjs")),
                mode,
                str(source),
                str(destination),
                str(self.home / "node"),
            ],
            capture_output=True,
            text=True,
            encoding="utf-8",
            timeout=60,
        )
        if result.returncode or not destination.is_file():
            raise ValueError(f"Native codec failed: {result.stderr[:1500]}")

    def decode(self, source: Path) -> etree._ElementTree:
        with TemporaryDirectory(prefix="bifrost-native-decode-") as directory:
            xml = Path(directory) / "decoded.xml"
            self._call("decode", source, xml)
            tree = parse_native_xml(xml.read_bytes())
        if tree.getroot().tag != "MSMElectronicsWorkbench" or tree.getroot().get("Schema") != "67":
            raise ValueError("Only the verified Multisim schema 67 is supported.")
        return tree

    def encode(self, tree: etree._ElementTree, output: Path) -> None:
        with TemporaryDirectory(prefix="bifrost-native-encode-") as directory:
            source = Path(directory) / "native.xml"
            source.write_bytes(native_xml_bytes(tree))
            self._call("encode", source, output)
